ambry-rest/src/main/java/com.github.ambry.rest/NettySslFactory.java

```diff
@@ -0,0 +1,198 @@
+ * instead of JDK's SSL implementation when the netty-tcnative library is loaded. OpenSSL shows
+ * significant performance enhancements over the JDK implementation.
+ */
+public class NettySslFactory implements SSLFactory {
+  private final SslContext nettyServerSslContext;
+  private final SslContext nettyClientSslContext;
+  private final SSLContext jdkSslContext;
+  private final String endpointIdentification;
+
+  /**
+   * Instantiate a {@link NettySslFactory} from a config.
+   * @param sslConfig the {@link SSLConfig} to use.
+   * @throws GeneralSecurityException
+   * @throws IOException
+   */
+  public NettySslFactory(SSLConfig sslConfig) throws GeneralSecurityException, IOException {
+    nettyServerSslContext = getServerSslContext(sslConfig);
+    nettyClientSslContext = getClientSslContext(sslConfig);
+    // Netty's OpenSsl based implementation does not use the JDK SSLContext so we have to fall back to the JDK based
+    // factory to support this method.
+    jdkSslContext = new JdkSslFactory(sslConfig).getSSLContext();
+
+    this.endpointIdentification =
+        sslConfig.sslEndpointIdentificationAlgorithm.isEmpty() ? null : sslConfig.sslEndpointIdentificationAlgorithm;
+  }
+
+  @Override
+  public SSLEngine createSSLEngine(String peerHost, int peerPort, Mode mode) {
+    SslContext context = mode == Mode.CLIENT ? nettyClientSslContext : nettyServerSslContext;
+    SSLEngine sslEngine = context.newEngine(ByteBufAllocator.DEFAULT, peerHost, peerPort);
+
+    if (mode == Mode.CLIENT) {
+      SSLParameters sslParams = sslEngine.getSSLParameters();
+      sslParams.setEndpointIdentificationAlgorithm(endpointIdentification);
+      sslEngine.setSSLParameters(sslParams);
+    }
+    return sslEngine;
+  }
+
+  @Override
+  public SSLContext getSSLContext() {
+    return jdkSslContext;
+  }
+
+  /**
+   * @param config the {@link SSLConfig}
+   * @return a configured {@link SslContext} object for a client.
+   * @throws GeneralSecurityException
+   * @throws IOException
+   */
+  private static SslContext getServerSslContext(SSLConfig config) throws GeneralSecurityException, IOException {
+    return SslContextBuilder.forServer(getKeyManagerFactory(config))
+        .trustManager(getTrustManagerFactory(config))
+        .sslProvider(SslProvider.JDK)
+        .ciphers(getCipherSuites(config))
+        .protocols(getEnabledProtocols(config))
+        .clientAuth(ClientAuth.REQUIRE)
+        .build();
+  }
+
+  /**
+   * @param config the {@link SSLConfig}
+   * @return a configured {@link SslContext} object for a server.
+   * @throws GeneralSecurityException
+   * @throws IOException
+   */
+  private static SslContext getClientSslContext(SSLConfig config) throws GeneralSecurityException, IOException {
+    return SslContextBuilder.forClient()
+        .keyManager(getKeyManagerFactory(config))
+        .trustManager(getTrustManagerFactory(config))
+        .sslProvider(SslProvider.OPENSSL)
```

Comment:
this is different from the one for server (JDK) - intended?

Reply:
Is there some logging that prints what it selected? I think that might be useful (since you say it is contingent on the libraries being present)

Reply:
We will now log while instantiating the `NettySslFactory`:
```
[2018-05-15 14:01:14,494] INFO Using OPENSSL provider for server SslContext (com.github.ambry.rest.NettySslFactory)
[2018-05-15 14:01:14,786] INFO Using OPENSSL provider for client SslContext (com.github.ambry.rest.NettySslFactory)
```